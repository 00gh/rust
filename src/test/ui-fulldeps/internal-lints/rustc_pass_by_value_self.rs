// compile-flags: -Z unstable-options
// NOTE: This test doesn't actually require `fulldeps`
// so we could instead use it as a `ui` test.
//
// Considering that all other `internal-lints` are tested here
// this seems like the cleaner solution though.
#![feature(rustc_attrs)]
#![deny(rustc::pass_by_value)]
#![allow(unused)]

#[rustc_diagnostic_item = "TyCtxt"]
#[rustc_pass_by_value]
struct TyCtxt<'tcx> {
    inner: &'tcx (),
}

impl<'tcx> TyCtxt<'tcx> {
    fn by_value(self) {} // OK
    fn by_ref(&self) {} //~ ERROR passing `TyCtxt<'tcx>` by reference
}

struct TyS<'tcx> {
    inner: &'tcx (),
}

#[rustc_diagnostic_item = "Ty"]
#[rustc_pass_by_value]
type Ty<'tcx> = &'tcx TyS<'tcx>;

impl<'tcx> TyS<'tcx> {
    fn by_value(self: Ty<'tcx>) {}
    fn by_ref(self: &Ty<'tcx>) {} //~ ERROR passing `Ty<'tcx>` by reference
}

fn main() {}
