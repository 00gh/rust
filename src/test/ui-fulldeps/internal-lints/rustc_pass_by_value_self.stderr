error: passing `TyCtxt<'tcx>` by reference
  --> $DIR/rustc_pass_by_value_self.rs:18:15
   |
LL |     fn by_ref(&self) {}
   |               ^^^^^ help: try passing by value: `TyCtxt<'tcx>`
   |
note: the lint level is defined here
  --> $DIR/rustc_pass_by_value_self.rs:8:9
   |
LL | #![deny(rustc::pass_by_value)]
   |         ^^^^^^^^^^^^^^^^^^^^

error: passing `Ty<'tcx>` by reference
  --> $DIR/rustc_pass_by_value_self.rs:30:21
   |
LL |     fn by_ref(self: &Ty<'tcx>) {}
   |                     ^^^^^^^^^ help: try passing by value: `Ty<'tcx>`

error: aborting due to 2 previous errors

